### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by eric-s-s**
first try at a test suite.

### CODE REVIEW

File: lim_test.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        self.basic2.remove(1, self.tuples)
+        self.basic2.remove(1, self.other_tuples)
+        self.basic2.remove(1, self.tuples)
+        self.basic2.remove(1, self.other_tuples)
+        self.assertEqual(self.basic2.frequency_all(), self.other_tuples)
+
+        self.basic1.remove(1, self.other_tuples)
+        self.assertEqual(self.basic1.frequency_all(), [(0, 1)])
+
+SUITE = unittest.TestLoader().loadTestsFromTestCase(TestLIMFunc)
```
Comment (zdsbs) [inline lim_test.py]: Kill all this suite stuff. You should just be running:

```
if __name__ == '__main__':
    unittest.main()
```

Then to execute run something like:

```
python -m unittest discover -p "*test.py"
```

or:

```
python lim_test.py
```

File: lim_test.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
+'''tests for the longintmath.py module'''
+import unittest
+import longintmath as lim
+
+class TestLIMFunc(unittest.TestCase):
+    '''tests for the long_int_(mathfunction) functions'''
+    def setUp(self):
+        self.long_big = 10**1000
```
Comment (zdsbs) [inline lim_test.py]: you can declare these at a higher scope and "mark" them as constants:

```
LONG_BIG = 10**1000
...
```

Comment (zdsbs) [inline lim_test.py]: since these are not mutated you do not need them in the setup and tear down. Those are only for mutable state.

File: lim_test.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        self.float_big = 1.0e+300
+        self.float_sm = 1.0e-100
+
+    def tearDown(self):
+        del self.long_big
+        del self.float_big
+        del self.float_sm
+
+    def error_tester(self, result, expected):
```
Comment (zdsbs) [inline lim_test.py]: Perhaps I'm wrong, but it looks like you should be using assertAlmostEqual(a, b) which allows you to give the number of places and a delta.

If you keep this method the name should begin with assert and say what it's asserting (ie how it is different)

File: lim_test.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        '''use the inverse error to avoid trouble with overflow errors.'''
+        allowed_error = 1e-8
+        if result - expected == 0:
+            inv_error = float('inf')
+        else:
+            inv_error = abs(expected // (result - expected))
+        self.assertGreater(inv_error, 1/allowed_error)
+
+    def test_li_div(self):
```
Comment (zdsbs) [inline lim_test.py]: Think one assertion per test http://blog.jayfields.com/2007/06/testing-one-assertion-per-test.html

File: lim_test.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
+'''tests for the longintmath.py module'''
```
Comment (zdsbs) [inline lim_test.py]: This file name should be longintmath_test.py don't be lazy with abbreviations.

File: lim_test.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+import unittest
+import longintmath as lim
+
+class TestLIMFunc(unittest.TestCase):
```
Comment (zdsbs) [inline lim_test.py]: I'm un-convinced you need multiple test class functions - But give them better non-abbrievated names and you can kill your comments about why they exist

File: lim_test.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        else:
+            inv_error = abs(expected // (result - expected))
+        self.assertGreater(inv_error, 1/allowed_error)
+
+    def test_li_div(self):
+        '''test long_int_div'''
+        self.assertEqual(lim.long_int_div(self.float_big, self.long_big), 0)
+        self.assertEqual(lim.long_int_div(self.float_sm, self.float_big), 0)
+
+        result1 = lim.long_int_div(self.float_big, self.float_sm)
```
Comment (zdsbs) [inline lim_test.py]: what different cases are these testing? When you revisit this code in a year will you remember what result1-4 are representing? 

These should be test methods where the method name describes why it exists. Do not be afraid on long test method names, they're not real methods (no one calls them). They're glorified comments.

File: lim_test.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    def test_li_pow(self):
+        '''test long_int_pow'''
+        result1 = lim.long_int_pow(self.long_big, 2.5)
+        self.error_tester(result1, 10**2500)
+
+
+class TestLITable(unittest.TestCase):
+    '''tests for all functions on a LongIntTable that aren't add and remove'''
+    def setUp(self):
+        self.table_a = lim.LongIntTable({3:3, 1:1, 2:2})
```
Comment (zdsbs) [inline lim_test.py]: why do you have these different tables? Can you justify their existence? What do they represent?

Why are there duplicates? Etc.

File: lim_test.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+        del self.table_b
+        del self.table_b1
+
+        del self.table_c
+
+        del self.empty_table_a
+        del self.empty_table_b
+
+    def test_values(self):
```
Comment (zdsbs) [inline lim_test.py]: test_values(self):
test_values_min(self):
test_values_max(self):
test_values_range(self):

test_values_emptyTable(self):
test_values_min_emptyTable(self):
test_values_max_emptyTable(self):
test_values_range_emptyTable(self):

Comment (zdsbs) [inline lim_test.py]: test only what you need to, to validate things are functioning correctly and name things in a way the help you understand if you're missing an edge case.

File: lim_test.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+
+        self.assertIsNone(self.empty_table_a.values_min())
+        self.assertIsNone(self.empty_table_a.values_max())
+        self.assertEqual(self.table_a.values_min(), 1)
+        self.assertEqual(self.table_a.values_max(), 3)
+
+        self.assertEqual(self.table_a.values_range(), (1, 3))
+        self.assertEqual(self.empty_table_b.values_range(), (None, None))
+
+    def test_frequency(self):
```
Comment (zdsbs) [inline lim_test.py]: same as values tests

File: lim_test.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        '''all the frequency_something functions do what they should'''
+        #confirms frequency function returns appropriate value or zero
+        self.assertEqual(self.table_a.frequency(2), (2, 2))
+        self.assertEqual(self.table_a.frequency(5), (5, 0))
```
Comment (zdsbs) [inline lim_test.py]: good additional test case, verifying what happens when the value does not exist. This is it's own test method.
test_frequency_when_out_of_range (or something like that)

File: lim_test.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+        self.assertEqual(self.table_a.frequency_range(0, 5),
+                         [(0, 0), (1, 1), (2, 2), (3, 3), (4, 0)])
+        #confirm freq all sorts and removes zero frequency
```
Comment (zdsbs) [inline lim_test.py]: your comment is a hint to yourself that you have a different method.

File: lim_test.py [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+
+    def test_stddev(self):
+        '''test the stddev() function'''
+        with self.assertRaises(ZeroDivisionError):
+            self.empty_table_a.stddev()
+        #confirms stddev computes correctly for numbers below cutoff for //
+        small_test = lim.LongIntTable(dict((x, 2) for x in range(1, 6)))
+        stddev = ((2*2**2 + 2*1)/5.)**0.5
+        self.assertEqual(small_test.stddev(), round(stddev, 4))
+        small_test = None
```
Comment (zdsbs) [inline lim_test.py]: what are you protecting against here?

File: lim_test.py [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+        #confirms stddev computes correctly for numbers above cutoff for //
+        large_test = lim.LongIntTable(dict((x, 10**100) for x in range(1, 6)))
+        self.assertEqual(large_test.stddev(), round(stddev, 4))
+        large_test = None
+        stddev = None
+    def test_merge_update(self):
```
Comment (zdsbs) [inline lim_test.py]: This test is too dependent on the state change in the previous section, you want to avoid that. This is yet another reason for separating these into different test method.

File: lim_test.py [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+        self.other_tuples = [(1, 1), (3, 1)]
+    def tearDown(self):
+        del self.basic1
+        del self.basic2
+        del self.table_a
+        del self.table_a1
+        del self.tuples
+        del self.other_tuples
+
+    def test_error_raising(self):
```
Comment (zdsbs) [inline lim_test.py]: of the tests that you have multiple assertions this is the only one that I might be convinced is okay. It tests a single function and shows you all the error cases around that function. It's concise.

File: lim_test.py [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+        '''make sure errors are raised and table doesn't mutate'''
+        self.assertRaisesRegexp(ValueError, 'cannot add an empty list',
+                                self.basic1.add, 1, [(1, 0)])
+        self.assertRaisesRegexp(ValueError, 'times must be a positive int',
+                                self.basic1.add, 0, [(1, 1)])
+        self.assertRaisesRegexp(ValueError, 'frequencies may not be negative',
+                                self.basic1.add, 1, [(1, -1)])
+        self.assertEqual(self.basic1.frequency_all(), [(0, 1)])
+    def test_order_zeros(self):
```
Comment (zdsbs) [inline lim_test.py]: I'm not sure why you have two examples here. What are you trying to show besides different entry points into the API?

File: lim_test.py [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+
+        self.table_a.add(1, [(1, 1), (2, 0), (3, 5)])
+        self.table_a1.add(1, [(3, 5), (1, 1)])
+        self.assertEqual(self.table_a.frequency_all(),
+                         self.table_a1.frequency_all())
+    def test_add(self):
+        '''test the cases for add'''
+        #add works with the identity table
+        self.basic1.add(1, self.tuples)
+        self.basic2.add(1, self.tuples)
```
Comment (zdsbs) [inline lim_test.py]: why are you adding to basic2 here?

File: lim_test.py [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+        self.assertEqual(self.table_a.frequency_all(),
+                         self.table_a1.frequency_all())
+
+        self.table_a.add(1, [(1, 1), (2, 0), (3, 5)])
+        self.table_a1.add(1, [(3, 5), (1, 1)])
+        self.assertEqual(self.table_a.frequency_all(),
+                         self.table_a1.frequency_all())
+    def test_add(self):
+        '''test the cases for add'''
+        #add works with the identity table
```
Comment (zdsbs) [inline lim_test.py]: is basic one the identity table? because that is a much better name.

File: lim_test.py [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+        self.basic1.add(1, self.tuples)
+        self.basic2.add(1, self.tuples)
+        self.assertEqual(self.basic1.frequency_all(), self.tuples)
+
+        #confirm when _fastest converts to tuples that it works
```
Comment (zdsbs) [inline lim_test.py]: I don't understand this. Is _fastest an internal method? If so how do you know you're hitting it?

File: lim_test.py [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+        self.table_a.add(1, [(val, 5) for val in range(1, 6)])
+        self.table_a1.add(1, [(val, 5) for val in range(5, 0, -1)])
+        self.assertEqual(self.table_a.frequency_all(),
+                         self.table_a1.frequency_all())
+
+        self.table_a.add(1, [(1, 1), (2, 0), (3, 5)])
+        self.table_a1.add(1, [(3, 5), (1, 1)])
+        self.assertEqual(self.table_a.frequency_all(),
+                         self.table_a1.frequency_all())
+    def test_add(self):
```
Comment (zdsbs) [inline lim_test.py]: In general for this method the question I find my self asking is, what are you trying to show with these different cases? Why do you need them to be so complicated?

File: lim_test.py [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+        empty_table.add(5, self.tuples)
+        self.assertEqual(empty_table.frequency_all(), [])
+        del empty_table
+
+    def test_remove(self):
+        '''test the remove function'''
+        self.basic1.add(2, self.tuples)
+        self.basic2.add(5, self.tuples)
+        self.basic2.remove(3, self.tuples)
+        self.assertEqual(self.basic1.frequency_all(), self.basic2.frequency_all())
```
Comment (zdsbs) [inline lim_test.py]: This feels okay to me.

File: lim_test.py [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+
+        self.basic1.add(1, self.other_tuples)
```
Comment (zdsbs) [inline lim_test.py]: I think this depends on the knowledge of what is contained in tuples and other_tuples, that's not good.

File: lim_test.py [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+        self.basic1.remove(2, self.tuples)
+        self.assertEqual(self.basic1.frequency_all(), self.other_tuples)
+
+        self.basic2.add(3, self.other_tuples)
```
Comment (zdsbs) [inline lim_test.py]: The rest of these are too complicated. I can't keep track of the state in basic2 or basic1 at this point... I'm lost. 

Simplify these if they should exist.

File: lim_test.py (hunk 3 above)
Comment (eric-s-s) [inline lim_test.py]: yup.  i tried assertAlmostEqual, but it has two problems.   1 - onlyworks floats.  overflowError for longs.  2 - it's asserting from absolute decimal.  so 3e300 -  2.999999999999e300 is 1e292 and it would only be true is the diff were <1e-7. (similarly, can't float div longs)

you mean like assertWithinError ?

File: lim_test.py (hunk 6 above)
Comment (eric-s-s) [inline lim_test.py]: i'm not sure about the actual function of the classes in a test suite.  i just took a stab.  is it usually one class per module?

Comment (zdsbs) [inline lim_test.py]: I think that it's okay usually in my world it is, but this is a judgement call - I'm not mad about it :D
